### PR TITLE
Improved Clifford synthesis method 

### PR DESCRIPTION
**Pull request by ShellyGarion**
### Summary
Add an improved Clifford synthesis method based on Bravyi et. al. greedy Clifford compiler.

### CODE REVIEW

**Comment by ShellyGarion** [comment]
@chriseclectic - the PR is ready for review - thanks!

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 1]
```diff
@@ -413,3 +418,290 @@ def _set_row_z_zero(clifford, circuit, qubit):
...
+# ---------------------------------------------------------------------
+
+# Global arrays of the 16 pairs of Pauli operators
+# divided into 5 equivalence classes under the action of single-qubit Cliffords
+A_class = [[[False, True], [True, True]],
+           [[False, True], [True, False]],
+           [[True, True], [False, True]],
+           [[True, True], [True, False]],
+           [[True, False], [False, True]],
+           [[True, False], [True, True]]]  # ['XY', 'XZ', 'YX', 'YZ', 'ZX', 'ZY']
```
Comment (chriseclectic) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:516]: Nit: you could comment the pauli on each line

**Review by chriseclectic — CHANGES_REQUESTED** [review]
Looks good, I just have a couple of suggestions for documentation

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 2]
```diff
@@ -43,13 +45,16 @@ def decompose_clifford(clifford):
         2. S. Aaronson, D. Gottesman, *Improved Simulation of Stabilizer Circuits*,
            Phys. Rev. A 70, 052328 (2004).
            `arXiv:quant-ph/0406196 <https://arxiv.org/abs/quant-ph/0406196>`_
+
+        3. Sergey Bravyi, Shaohan Hu, Dmitri Maslov, Ruslan Shaydulin,
+           *Clifford Circuit Optimization with Templates and Symbolic Pauli Gates*
```
[the remark this answers is not in the document]
Comment (ShellyGarion) [reply qiskit/quantum_info/synthesis/clifford_decompose.py:51]: there is no link yet.

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 3]
```diff
@@ -19,18 +19,21 @@
...
 
     For N <= 3 qubits this is based on optimal CX cost decomposition
     from reference [1]. For N > 3 qubits this is done using the general
-    non-optimal compilation routine from reference [2].
+    non-optimal greedy compilation routine from reference [3],
+    which yields better CX cost compared to [2].
 
     Args:
         clifford (Clifford): a clifford operator.
+        method (str): a synthesis method ('AG' or 'greedy')
```
Comment (chriseclectic) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:36]:
```suggestion
        method (str): Optional, a synthesis method ('AG' or 'greedy'). If set this overrides optimal decomposition for N <=3 qubits. 
```

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 4]
```diff
@@ -19,18 +19,21 @@
...
 
-def decompose_clifford(clifford):
+def decompose_clifford(clifford, method=None):
     """Decompose a Clifford operator into a QuantumCircuit.
 
     For N <= 3 qubits this is based on optimal CX cost decomposition
     from reference [1]. For N > 3 qubits this is done using the general
-    non-optimal compilation routine from reference [2].
+    non-optimal greedy compilation routine from reference [3],
+    which yields better CX cost compared to [2].
```
Comment (chriseclectic) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:32]:
```suggestion
    which typically yields better CX cost compared to the AG method [2].
```

**Review by chriseclectic — COMMENTED** [review]
Could more minor doc changes

**Comment by ajavadia** [comment]
can you add a reno releasenote please

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # This code is part of Qiskit.
 #
-# (C) Copyright IBM 2017, 2020
+# (C) Copyright IBM 2017, 2020, 2021
```
Comment (1ucian0) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:3]:
```suggestion
# (C) Copyright IBM 2017, 2021
```

**Review by chriseclectic — APPROVED** [review]

**Review by chriseclectic — COMMENTED** [review]
LGTM, just needs a release note

**Comment by ShellyGarion** [comment]
release notes are included: 
https://github.com/Qiskit/qiskit-terra/pull/5779/files#diff-09528904c5b6d193ec7eef03986f8ef40e5de9f37ba13164bcb1203e32b664a2

**Pull request merged**